logic/transactions/pool.js

```diff
@@ -0,0 +1,705 @@
+	var poolList = ['unverified','pending','ready'];
+
+	[pool.unverified, pool.verified.pending, pool.verified.ready].forEach(function (list, index) {
+		if (__private.delete(transaction.id, list)) {
+			clearedList.push(poolList[index]);
+		}
+	});
+	if (clearedList.length > 0) {
+		if (clearedList.length > 1) {
+			library.logger.debug(['Cleared duplicated tx in pool list:', clearedList, 'txid:', transaction.id].join(' '));
+		}
+	}
+	return clearedList;	
+};
+
+/**
+ * Pulls transactions from unverified, perform verifications and if successful
+ * push to either verified.pending (when tx is multisign or timestamp is in 
+ * future) or verified.ready otherwise.
+ * @implements {__private.delete}
+ * @implements {processVerifyTransaction}
+ * @implements {removeUnconfirmedTransaction}
+ * @param {function} cb
+ * @return {setImmediateCallback} err | cb
+ */
+TxPool.prototype.processPool = function (cb) {
+	async.series({
+		expireTransactions: function (seriesCb) {
+			__private.expireTransactions(function (err) {
+				if (err) {
+					library.logger.log('Transaction expiry timer', err);
+				}
+				return setImmediate(seriesCb);
+			});
+		},
+		processUnverified: function (seriesCb) {
+			async.eachSeries(pool.unverified.transactions, function (transaction, eachSeriesCb) {
+				__private.delete(transaction.id, pool.unverified);
+				__private.processUnverifiedTransaction(transaction, true, function (err, sender) {
+					if (err) {
+						library.logger.error('Failed to process unverified transaction: ' + transaction.id, err);
+						return setImmediate(eachSeriesCb);
+					}
+					self.checkBalance(transaction, sender, function (err) {
+						if (err) {
+							library.logger.error('Failed to check balance transaction: ' + transaction.id, err);
+						}
+						transaction.receivedAt = new Date();
+						if (transaction.type === transactionTypes.MULTI || Array.isArray(transaction.signatures || transaction.receivedAt < transaction.timestamp)) {
+							__private.add(transaction, pool.verified.pending, eachSeriesCb);
+						} else {
+							// check transaction and if ok add to verified.ready
+							__private.add(transaction, pool.verified.ready, eachSeriesCb);
+						}
+					});
+				});
+			}, function (err) {
+				return setImmediate(seriesCb, err);
+			});
+		},
+		processPending: function (seriesCb) {
+			// process pool.verified.pending (multisig txs signs), and take care 
+			// about moving transactions from `verified.pending` to `verified.ready`
+			async.eachSeries(pool.verified.pending.transactions, function (transaction, eachSeriesCb) {
+				__private.delete(transaction.id, pool.verified.pending);
+				__private.add(transaction, pool.verified.ready, eachSeriesCb);
+			}, function (err) {
+				return setImmediate(seriesCb, err);
+			});
+		},
+		reindexQueues: function (seriesCb) {
+			if (self.processed > 1000) {
+				__private.reindexQueues();
+				self.processed = 0;
+			}
+		}
+	}, function (err) {
+		return setImmediate(cb, err);
```

Comment:
if there was error during processing - we should store transaction (or just id) for a period of time to prevent accept same bad transaction again - maybe in `pool.invalid` and clear them after some time